haystack_experimental/components/builders/chat_prompt_builder.py

```diff
@@ -0,0 +1,341 @@
+    ```
+
+    #### New experimental prompt template
+    ```python
+    from haystack_experimental.components.builders import ChatPromptBuilder
+    from haystack.dataclasses.image_content import ImageContent
+
+    template = \"\"\"
+    {% message role="system" %}
+    You are a helpful assistant.
+    {% endmessage %}
+
+    {% message role="user" %}
+    Hello! I am {{user_name}}. What's the difference between the following images?
+    {% for image in images %}
+    {{ image | templatize_part }}
+    {% endfor %}
+    {% endmessage %}
+    \"\"\"
+
+    images = [ImageContent.from_file_path("apple.jpg"), ImageContent.from_file_path("orange.jpg")]
+
+    builder = ChatPromptBuilder(template=template)
+    builder.run(user_name="John", images=images)
+    ```
+    """
+
+    def __init__(
+        self,
+        template: Optional[Union[List[ChatMessage], str]] = None,
+        required_variables: Optional[Union[List[str], Literal["*"]]] = None,
+        variables: Optional[List[str]] = None,
+    ):
+        """
+        Constructs a ChatPromptBuilder component.
+
+        :param template:
+            A list of `ChatMessage` objects or a string template. The component looks for Jinja2 template syntax and
+            renders the prompt with the provided variables. Provide the template in either
+            the `init` method` or the `run` method.
+        :param required_variables:
+            List variables that must be provided as input to ChatPromptBuilder.
+            If a variable listed as required is not provided, an exception is raised.
+            If set to "*", all variables found in the prompt are required. Optional.
+        :param variables:
+            List input variables to use in prompt templates instead of the ones inferred from the
+            `template` parameter. For example, to use more variables during prompt engineering than the ones present
+            in the default template, you can provide them here.
+        """
+        self._variables = variables
+        self._required_variables = required_variables
+        # self.required_variables = required_variables or []
+        self.template = template
+        # variables = variables or []
+
+        self._env = SandboxedEnvironment(extensions=[ChatMessageExtension])
+        self._env.filters["templatize_part"] = templatize_part
+        if arrow_import.is_successful():
+            self._env.add_extension(Jinja2TimeExtension)
+
+        extracted_variables = []
+        if template and not variables:
+            if isinstance(template, list):
+                for message in template:
+                    if message.is_from(ChatRole.USER) or message.is_from(ChatRole.SYSTEM):
+                        # infer variables from template
+                        if message.text is None:
+                            raise ValueError(f"The provided ChatMessage has no text. ChatMessage: {message}")
```

Comment:
Let's add a bit more of a call to action here. So say that text always need to be added to the ChatMessage

Reply:
I improved the error message. BTW, this is quite unlikely to happen if users use `ChatMessage.from_user` etc.. It was added to make mypy happy.